File: Algorithms/golang-sorting/sort/selectionsort/selectionsort.go
package selectionsort

import "heartiger.com/algorithms/utility"

// In every outer loop we assume current element is the smallest
// among elements after it. If we found any element that's smaller than
// it, update the smallest index. Finally swap the smallest element to the front.
// Complexity is O(n2)
func SelectionSort(nums []int) {
	for i := 0; i < len(nums); i++ {
		minIndex := i
		for j := i+1; j < len(nums); j++ {
			if nums[j] < nums[minIndex] {
				minIndex = j
			}
		}
		if minIndex != i {
			utility.Swap(nums, minIndex, i)
		}
	}
}

File: Algorithms/Sorting/src/sorters/SelectionSort.java
package sorters;

import utility.Sortable;
import utility.Utility;

public class SelectionSort implements Sortable {

    private int[] nums;

    public SelectionSort(){}
    public SelectionSort(int[] nums){this.nums = nums;}


    @Override
    public int[] sort() {
        for(int i=0; i<nums.length; i++){
            for(int j=i+1; j<nums.length; j++)
                if (nums[j] < nums[i])
                    Utility.swap(nums, i, j);
        }
        return nums;
    }

    @Override
    public void setArray(int[] nums) {
        this.nums = nums;
    }
}
